base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -472,99 +451,108 @@ fn get_db_encryption(
 
 fn get_cipher_for_db_encryption(
     passphrase: SafePassword,
-    passphrase_hash: Option<String>,
-    encryption_salt: Option<String>,
-    secret_seed: &Option<String>,
+    secondary_key_version: Option<String>,
+    secondary_key_salt: Option<String>,
+    encrypted_main_key: Option<String>,
     conn: &SqliteConnection,
 ) -> Result<XChaCha20Poly1305, WalletStorageError> {
-    let encryption_salt = match (passphrase_hash, encryption_salt) {
-        (None, None) => {
-            // Use the recommended OWASP parameters, which are not the default:
-            // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
-
-            // These are the parameters for the passphrase hash
-            let params_passphrase = argon2::Params::new(
-                46 * 1024, // m-cost: 46 MiB, converted to KiB
-                1,         // t-cost
-                1,         // p-cost
-                None,      // output length: default is fine for this use
-            )
-            .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
-
-            // Hash the passphrase to a PHC string for later verification
-            let passphrase_salt = SaltString::generate(&mut OsRng);
-            let passphrase_hash = argon2::Argon2::default()
-                .hash_password_customized(
+    // Set up the PBKDF used for secondary key derivation
+    // These parameters will likely change in the future, so we version them
+    // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
```

Comment:
Changing these parameters in production, might be extremely sensitive, as it will impact all wallet data encryption, across the network. On which basis, apart from possibly security breaches on the protocols, will these parameters change in the future ?

Reply:
Sources like [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#upgrading-the-work-factor) update their parameter recommendations over time to account for things like hardware advances. Fortunately, we can account for this by always supporting decryption of older versions, and by seamlessly updating by re-encrypting the main key.